source/extensions/filters/udp/dns_filter/dns_filter.cc

```diff
@@ -3,55 +3,59 @@
-  static constexpr std::chrono::seconds DEFAULT_RESOLVER_TTL{300};
-
   const auto& server_config = config.server_config();
 
-  // TODO(abaptiste): Read the external DataSource
-  if (server_config.has_inline_dns_table()) {
-    const auto& dns_table = server_config.inline_dns_table();
-    const size_t entries = dns_table.virtual_domains().size();
-
-    virtual_domains_.reserve(entries);
-    for (const auto& virtual_domain : dns_table.virtual_domains()) {
-      AddressConstPtrVec addrs{};
-      if (virtual_domain.endpoint().has_address_list()) {
-        const auto& address_list = virtual_domain.endpoint().address_list().address();
-        addrs.reserve(address_list.size());
-        // This will throw an exception if the configured_address string is malformed
-        for (const auto& configured_address : address_list) {
-          const auto ipaddr =
-              Network::Utility::parseInternetAddress(configured_address, 0 /* port */);
-          addrs.push_back(ipaddr);
-        }
+  envoy::data::dns::v3::DnsTable dns_table;
+  bool result = loadServerConfig(server_config, dns_table);
+  ENVOY_LOG(debug, "Loading DNS table from external file: {}", result ? "Success" : "Failure");
+
+  retry_count_ = dns_table.external_retry_count();
+
+  const size_t entries = dns_table.virtual_domains().size();
+  virtual_domains_.reserve(entries);
+  for (const auto& virtual_domain : dns_table.virtual_domains()) {
+    AddressConstPtrVec addrs{};
+    if (virtual_domain.endpoint().has_address_list()) {
+      const auto& address_list = virtual_domain.endpoint().address_list().address();
+      addrs.reserve(address_list.size());
+      // Creating the IP address will throw an exception if the address string is malformed
+      for (const auto& address : address_list) {
+        auto ipaddr = Network::Utility::parseInternetAddress(address, 0 /* port */);
+        addrs.push_back(std::move(ipaddr));
       }
-      virtual_domains_.emplace(virtual_domain.name(), std::move(addrs));
-      std::chrono::seconds ttl = virtual_domain.has_answer_ttl()
-                                     ? std::chrono::seconds(virtual_domain.answer_ttl().seconds())
-                                     : DEFAULT_RESOLVER_TTL;
-      domain_ttl_.emplace(virtual_domain.name(), ttl);
     }
+    virtual_domains_.emplace(virtual_domain.name(), std::move(addrs));
 
-    // Add known domains
-    known_suffixes_.reserve(dns_table.known_suffixes().size());
-    for (const auto& suffix : dns_table.known_suffixes()) {
-      auto matcher_ptr = std::make_unique<Matchers::StringMatcherImpl>(suffix);
-      known_suffixes_.push_back(std::move(matcher_ptr));
-    }
+    std::chrono::seconds ttl = virtual_domain.has_answer_ttl()
+                                   ? std::chrono::seconds(virtual_domain.answer_ttl().seconds())
+                                   : DEFAULT_RESOLVER_TTL;
+    domain_ttl_.emplace(virtual_domain.name(), ttl);
+  }
+
+  // Add known domain suffixes
```

Comment:
The virtual domains must match an existing suffix.   For example you could add `jp.com`to the known suffix list, and then define a unique set of answers for virtual domains such as `raptor.jp.com` or `t-rex.jp.com`.   Since both domains match on the common suffix, then the filter will respond to the query